ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/jdbc/JDBCStore.java

```diff
@@ -201,65 +204,79 @@ public JDBCStore(ObjectStoreEnvironmentBean jdbcStoreEnvironmentBean) throws Obj
             impleTableName = tablePrefix + impleTableName;
         }
         tableName = impleTableName;
-        key = connectionDetails + tableName;
-
+        key = ds == null ? connectionDetails + tableName : ds + tableName;
         _theImple = imples.get(key);
         _storeName = storeNames.get(key);
+
         if (_theImple == null) {
-            try {
+            initImple(ds, connectionDetails, key);
+        }
+    }
+
+    private void initImple(DataSource dataSource, String connectionDetails, String key) throws ObjectStoreException {
+        try {
+            JDBCAccess jdbcAccess;
+
+            if (dataSource != null) {
+                // dataSource takes priority over connectionDetails
+                jdbcAccess = new DirectDataSourceJDBCAccess(dataSource);
+                jdbcAccess.initialise(null);
+            } else if (connectionDetails != null) {
                 StringTokenizer stringTokenizer = new StringTokenizer(connectionDetails, ";");
-                JDBCAccess jdbcAccess = (JDBCAccess) Class.forName(stringTokenizer.nextToken()).newInstance();
+                jdbcAccess = (JDBCAccess) Class.forName(stringTokenizer.nextToken()).newInstance();
                 jdbcAccess.initialise(stringTokenizer);
+            } else {
+                throw new ObjectStoreException("Invalid store configuration");
+            }
 
-                _storeName = jdbcAccess.getClass().getName() + ":" + tableName;
```

Comment:
Could you please group all formatting changes in a separate commit? Thanks :-)

Reply:
This one is not reformatting - I moved this code into its own method: `private void initImple(...`.